File: src-tauri/src/providers/adapters/factory.rs
//! Provider Adapter Factory
//! Creates the appropriate adapter based on provider name

use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::sync::Arc;

use super::gemini::GeminiAdapter;
use super::glm::GLMAdapter;
use super::kimi::KimiAdapter;
use super::minimax::MinimaxAdapter;
use super::modelscope::ModelScopeAdapter;
use super::nvidia::NvidiaNimAdapter;
use super::openai::{OpenAIAdapter, OpenAICompatibilityAdapter};
use super::siliconflow::SiliconFlowAdapter;
use super::traits::ProviderAdapter;

// Cache for adapter instances
static ADAPTER_CACHE: Lazy<HashMap<String, Arc<dyn ProviderAdapter>>> = Lazy::new(|| {
    let mut map = HashMap::new();

    // OpenAI and compatible
    map.insert(
        "openai".to_string(),
        Arc::new(OpenAIAdapter::new()) as Arc<dyn ProviderAdapter>,
    );
    map.insert(
        "openai_compatibility".to_string(),
        Arc::new(OpenAICompatibilityAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // SiliconFlow
    map.insert(
        "siliconflow".to_string(),
        Arc::new(SiliconFlowAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // GLM (Zhipu AI)
    map.insert(
        "glm".to_string(),
        Arc::new(GLMAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // Kimi (Moonshot AI)
    map.insert(
        "kimi".to_string(),
        Arc::new(KimiAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // ModelScope
    map.insert(
        "modelscope".to_string(),
        Arc::new(ModelScopeAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // Gemini (Google)
    map.insert(
        "gemini".to_string(),
        Arc::new(GeminiAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // Nvidia NIM
    map.insert(
        "nvidia".to_string(),
        Arc::new(NvidiaNimAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    // MiniMax
    map.insert(
        "minimax".to_string(),
        Arc::new(MinimaxAdapter::new()) as Arc<dyn ProviderAdapter>,
    );

    map
});

/// Get provider adapter instance
/// Returns a cached adapter for the specified provider
pub fn get_provider_adapter(provider: &str) -> Arc<dyn ProviderAdapter> {
    // Return cached instance if available
    if let Some(adapter) = ADAPTER_CACHE.get(provider) {
        return adapter.clone();
    }

    // Fallback to OpenAI adapter for unknown providers
    // (assumes OpenAI-compatible API)
    ADAPTER_CACHE.get("openai").unwrap().clone()
}

/// Check if provider is supported
pub fn is_provider_supported(provider: &str) -> bool {
    ADAPTER_CACHE.contains_key(provider)
}

/// List all supported providers
pub fn supported_providers() -> Vec<&'static str> {
    ADAPTER_CACHE.keys().map(|s| s.as_str()).collect()
}

File: backend-python/src/providers/factory.py
"""
Provider adapter factory.
Creates the appropriate adapter based on provider name.
"""


from .base import BaseProviderAdapter
from .openai import OpenAIAdapter
from .other_providers import (
    DeepSeekAdapter,
    GeminiAdapter,
    GLMAdapter,
    HuggingFaceAdapter,
    KimiAdapter,
    LiteLLMOpenAIAdapter,
    MinimaxAdapter,
    ModelScopeAdapter,
    NvidiaAdapter,
    OpenRouterAdapter,
    SiliconFlowAdapter,
)

# Cache adapter instances for reuse
_adapter_cache: dict[str, BaseProviderAdapter] = {}


def get_provider_adapter(
    provider: str,
) -> BaseProviderAdapter:
    """
    Get provider adapter instance.

    Args:
        provider: Provider name (openai, gemini, siliconflow, etc.)

    Returns:
        Provider adapter instance
    """
    # Return cached instance if available
    if provider in _adapter_cache:
        return _adapter_cache[provider]

    # Create new adapter instance
    adapter: BaseProviderAdapter

    match provider:
        case "openai" | "openai_compatibility":
            adapter = OpenAIAdapter()
        case "siliconflow":
            adapter = SiliconFlowAdapter()
        case "openrouter":
            adapter = OpenRouterAdapter()
        case "litellm_openai":
            adapter = LiteLLMOpenAIAdapter()
        case "huggingface":
            adapter = HuggingFaceAdapter()
        case "kimi":
            adapter = KimiAdapter()
        case "glm":
            adapter = GLMAdapter()
        case "deepseek":
            adapter = DeepSeekAdapter()
        case "volcengine":
            adapter = GLMAdapter()
        case "modelscope":
            adapter = ModelScopeAdapter()
        case "gemini":
            adapter = GeminiAdapter()
        case "nvidia":
            adapter = NvidiaAdapter()
        case "minimax":
            adapter = MinimaxAdapter()
        case _:
            # Fallback to OpenAI adapter for unknown providers
            # (assumes OpenAI-compatible API)
            import warnings
            warnings.warn(f"Unknown provider: {provider}, using OpenAI adapter as fallback")
            adapter = OpenAIAdapter()

    # Cache for future use
    _adapter_cache[provider] = adapter
    return adapter


def is_provider_supported(provider: str) -> bool:
    """Check if provider is supported."""
    return provider in [
        "openai",
        "openai_compatibility",
        "siliconflow",
        "openrouter",
        "litellm_openai",
        "huggingface",
        "kimi",
        "glm",
        "deepseek",
        "volcengine",
        "modelscope",
        "gemini",
        "nvidia",
        "minimax",
    ]


# Export all supported providers
SUPPORTED_PROVIDERS: list[str] = [
    "openai",
    "openai_compatibility",
    "siliconflow",
    "openrouter",
    "litellm_openai",
    "huggingface",
    "kimi",
    "glm",
    "deepseek",
    "volcengine",
    "modelscope",
    "gemini",
    "nvidia",
    "minimax",
]
